test/unit/master/test_cluster_master.py

```diff
@@ -213,6 +215,29 @@ def test_updating_slave_to_nonexistent_state_should_raise_bad_request_error(self
         with self.assertRaises(BadRequestError):
             master.handle_slave_state_update(slave, 'NONEXISTENT_STATE')
 
+    @genty_dataset (
+        slave_unresponsive=(True,1000,),
+        slave_dead=(False,60,),
+        slave_responsive=(True,60,),
+    )
+    def test_heartbeat_disconnects_unresponsive_slave(self, slave_alive, seconds_since_last_heartbeat):
+        master = ClusterMaster()
+        slave_url = "url"
+        slave = master.connect_slave(slave_url, 1)
+        slave = master.get_slave(int(slave['slave_id']))
+        last_heartbeat_time = datetime.now() - timedelta(seconds=seconds_since_last_heartbeat)
+
+        slave.is_alive = MagicMock(return_value=slave_alive)
+        slave.get_last_heartbeat_time = MagicMock(return_value=last_heartbeat_time)
+        slave.mark_dead = Mock()
```

Comment:
Do you need a real slave instance here? It's a bit surprising to see directly replacing methods on real objest with mocks via assignment and can lead to surprising results and stale tests (for example if someone renamed the mark_dead method but did not rename it here, your assertion about it never being called would pass trivially.) I believe the `patch` functionality does checks to guard against this.

More specifically, you should be able to patch out the `Slave` constructor so that you can inject a full mock Slave object that also asserts method calls match the real object api.